compiler/rustc_hir_analysis/src/collect/resolve_bound_vars.rs

```diff
@@ -2157,6 +2147,86 @@ impl<'a, 'tcx> BoundVarContext<'a, 'tcx> {
+                        | Scope::TraitRefBoundary { s }
+                        | Scope::LateBoundary { s, .. }
+                        | Scope::Opaque { s, .. } => {
+                            next_scope = Some(s);
+                            continue;
+                        }
+                        Scope::Root { opt_parent_item } => {
+                            if let Some(parent_id) = opt_parent_item {
+                                self.tcx.local_def_id_to_hir_id(parent_id)
+                            } else {
+                                continue;
+                            }
+                        }
+                    };
+                    let node = self.tcx.hir_node(hir_id);
+                    // If this is a `Self` bound in a trait, yield the trait itself.
+                    // Specifically, we don't need to look at any supertraits since
+                    // we already do that in `BoundVarContext::supertrait_hrtb_vars`.
+                    if let Res::SelfTyParam { trait_: _ } = expected_res
+                        && let hir::Node::Item(item) = node
+                        && let hir::ItemKind::Trait(..) = item.kind
+                    {
+                        // Yield the trait's def id. Supertraits will be
+                        // elaborated from that.
+                        yield item.owner_id.def_id.to_def_id();
+                    } else if let Some(generics) = node.generics() {
+                        for pred in generics.predicates {
+                            let hir::WherePredicate::BoundPredicate(pred) = pred else {
+                                continue;
+                            };
+                            let hir::TyKind::Path(hir::QPath::Resolved(None, bounded_path)) =
+                                pred.bounded_ty.kind
+                            else {
+                                continue;
+                            };
+                            // Match the expected res.
+                            if bounded_path.res != expected_res {
+                                continue;
+                            }
+                            for pred in pred.bounds {
+                                match pred {
+                                    hir::GenericBound::Trait(poly_trait_ref) => {
+                                        if let Some(def_id) =
+                                            poly_trait_ref.trait_ref.trait_def_id()
+                                        {
+                                            yield def_id;
+                                        }
+                                    }
+                                    hir::GenericBound::Outlives(_)
+                                    | hir::GenericBound::Use(_, _) => {}
+                                }
+                            }
+                        }
+                    }
+                }
+            },
+        )
+        .fuse()
+    }
```

Comment:
I wonder how hard it would be to reuse the method probing code we have in rustc_hir_typeck.

Reply:
I think it would be pretty difficult to do without inducing cycles, since the lowering code requires lowering trait refs.